li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java

```diff
@@ -392,39 +531,40 @@ public void wakeup() {
 
   // Intended for testing only
   LiKafkaConsumer<K, V> getPerClusterConsumer(ClusterDescriptor cluster) {
-    return _consumers.get(cluster);
-  }
-
-  private LiKafkaConsumer<K, V> getOrCreateConsumerForTopic(String topic) {
-    if (topic == null || topic.isEmpty()) {
-      throw new IllegalArgumentException("Topic cannot be null or empty");
-    }
-
-    // TODO: Handle nonexistent topics more elegantly with auto topic creation option
-    ClusterDescriptor cluster = _mdsClient.getClusterForTopic(_clientId, topic, _mdsRequestTimeoutMs);
-    if (cluster == null) {
-      throw new IllegalStateException("Topic " + topic + " not found in the metadata service");
+    // Get an immutable copy of the current set of consumers.
+    List<ClusterConsumerPair<K, V>> consumers = _consumers;
+    for (ClusterConsumerPair<K, V> entry : consumers) {
+      if (entry.getCluster().equals(cluster)) {
+        return entry.getConsumer();
+      }
     }
-
-    return getOrCreatePerClusterConsumer(_mdsClient.getClusterForTopic(_clientId, topic, _mdsRequestTimeoutMs));
+    return null;
   }
 
   // Returns null if the specified topic does not exist in the cluster group.
-  private LiKafkaConsumer<K, V> getOrCreatePerClusterConsumer(ClusterDescriptor cluster) {
+  private LiKafkaConsumer<K, V> getOrCreatePerClusterConsumer(List<ClusterConsumerPair<K, V>> consumers,
+      ClusterDescriptor cluster) {
     if (cluster == null) {
       throw new IllegalArgumentException("Cluster cannot be null");
     }
 
-    if (_consumers.containsKey(cluster)) {
-      return _consumers.get(cluster);
+    for (ClusterConsumerPair<K, V> entry : consumers) {
+      if (entry.getCluster().equals(cluster)) {
+        return entry.getConsumer();
+      }
     }
 
-    // Create per-cluster consumer config with the actual bootstrap URL of the physical cluster to connect to.
+    // Create per-cluster consumer config where the following cluster-specific properties:
+    //   - bootstrap.server - the actual bootstrap URL of the physical cluster to connect to
+    //   - max.poll.records - the property value set for the federated consumer / the number of clusters in this cluster group
+    //                        if the federated consumer property < the number of clusters in the group, set it to 1
+    //                        (poll() will make sure that it won't fetch more than the federated consumer property)
     Map<String, Object> configMap = _commonConsumerConfigs.originals();
     configMap.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, cluster.bootstrapURL());
-    _consumerBuilder.setConsumerConfig(configMap);
-    LiKafkaConsumer<K, V> newConsumer = _consumerBuilder.build();
-    _consumers.put(cluster, newConsumer);
+    int maxPollRecordsPerCluster = Math.max(_maxPollRecordsForFederatedConsumer / _numClustersToConnectTo, 1);
```

Comment:
in theory wont this require adjusting max poll records on existing (other) consumers? should we at least capture this as a TODO ?

Reply:
For now, it is okay because we support static assignment as well. A new assignment would close all existing consumers and recreate them. 

It'll be trickier with subscriptions, though. I'll add a TODO.